Feat create artifact repository and service

AlisaUrsu opened the pull request:
Created the repository and service for Artifact with methods based on the endpoints specified in the issues. Probably more methods will be added later.

AlisaUrsu committed:
feat: create repository and service for Artifact

I implemented all the operations mentioned in the issues, more might appear later on. I think it looks good now, maybe some constraints exception might be needed?

AlisaUrsu committed:
feat: add more methods for Artifact repository and service

CalinescuFlorin reviewed (changes requested):
Pare ok, doar că ai un query invalid in repository

Cred că ar trebui să fie ceva de genul:
```
SELECT Artifacts.* FROM Artifacts 
INNER JOIN Section ON Artifacts.section_id = Section.id
INNER JOIN Site ON Section.site_id = Site.id
INNER JOIN Site_Archaeologist ON Site.id = Site_Archaeologist.site_id
WHERE Site.id = :siteId AND Site_Archaeologist.archaeologist_id = :archaeologistId;
```
Nu stiu dacă face ceea ce voiai, te rog să verifici și să corectezi.

dunmer0 commented:
> Pare ok, doar că ai un query invalid in repository
> 
> Cred că ar trebui să fie ceva de genul:
> 
> ```
> SELECT Artifacts.* FROM Artifacts 
> INNER JOIN Section ON Artifacts.section_id = Section.id
> INNER JOIN Site ON Section.site_id = Site.id
> INNER JOIN Site_Archaeologist ON Site.id = Site_Archaeologist.site_id
> WHERE Site.id = :siteId AND Site_Archaeologist.archaeologist_id = :archaeologistId;
> ```
> 
> Nu stiu dacă face ceea ce voiai, te rog să verifici și să corectezi.



In mod normal in Query scrii JPQL un fel de SQL dar interpretat de Java Persistence si 
pentru a putea scrie native SQL treubie adugat la urma nativeQuery = true. 
**e.g: `@Query(value = "some native SQL query", nativeQuery = true)`**

AlisaUrsu committed:
fix: added a validator and exception class for service, a missing method and tests

The query works correctly, I just refactored it a bit. I added a new method which retrieves artifacts by a category (I suppose we will modify the type of category from string to enum once we know exactly the categories we'll work with later). I also tested every method in the service.

AlisaUrsu commented:
Am testat query-ul si functioneaza cum trebuie, doar i-am dat refactor putin sa arate mai clean. Am mai adaugat un validator pentru service-ul de Artifact si un exception class, inca o metoda care da retrieve la artefacte in functie de categorie (banuiesc ca o sa modificam category sa fie de tip enum in loc de string dupa ce o sa stim care sunt categoriile exacte). Am adaugat si teste pentru toate metodele din service.

CalinescuFlorin reviewed (approved):
looks good to me

CalinescuFlorin commented:
> categorie (banuiesc ca o sa modificam category sa fie de tip enum in loc de string dupa ce o sa stim care sunt categoriile exacte).

Gabi spunea ca nu vor fi categorii stabilite, ci vor fi introduse de user; momentan lasam asa, daca va fi cazul, facem tabela separata pt categorii
